File: src/main/java/de/tectoast/emolga/utils/RequestGetter.java
package de.tectoast.emolga.utils;

import com.google.api.services.sheets.v4.model.ExtendedValue;
import com.google.api.services.sheets.v4.model.Spreadsheet;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

@SuppressWarnings("UnusedReturnValue")
public class RequestGetter {
    private static final Logger logger = LoggerFactory.getLogger(RequestGetter.class);
    private final String sid;
    private final List<String> ranges = new LinkedList<>();

    public RequestGetter(String sid) {
        this.sid = sid;
    }

    public RequestGetter addRange(String range) {
        ranges.add(range);
        return this;
    }

    public List<List<List<String>>> execute() {
        try {
            Spreadsheet sh = Google.getSheetsService().spreadsheets().get(sid).setIncludeGridData(true).setRanges(ranges).execute();
            HashMap<String, AtomicInteger> map = new HashMap<>();
            List<List<List<String>>> ret = new ArrayList<>(ranges.size());
            for (String range : ranges) {
                logger.info("range = {}", range);
                String sheetname = range.split("!")[0];
                ret.add(sh.getSheets().stream().filter(s -> s.getProperties().getTitle().equals(sheetname)).findFirst().orElse(null).getData()
                        .get(map.computeIfAbsent(sheetname, i -> new AtomicInteger(-1)).incrementAndGet())
                        .getRowData().stream().map(rd -> rd.getValues().stream().filter(cd -> cd.getEffectiveValue() != null).map(cd -> {
                            ExtendedValue v = cd.getEffectiveValue();
                            if (v.getStringValue() != null) return v.getStringValue();
                            double d = v.getNumberValue();
                            if (d % 1.0 != 0)
                                return String.format("%s", d);
                            else
                                return String.format("%.0f", d);
                        }).collect(Collectors.toList())).collect(Collectors.toList()));
            }
            return ret;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: src/main/kotlin/de/tectoast/emolga/utils/RequestGetter.kt
package de.tectoast.emolga.utils

import com.google.api.services.sheets.v4.model.CellData
import de.tectoast.emolga.utils.Google.sheetsService
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

@Suppress("unused")
class RequestGetter(private val sid: String) {
    private val ranges: MutableList<String> = LinkedList()
    fun addRange(range: String): RequestGetter {
        ranges.add(range)
        return this
    }

    fun execute(): List<List<List<String>>> {
        val sh = sheetsService!!.spreadsheets()[sid].setIncludeGridData(true).setRanges(ranges)
            .execute() //Google.getSheetsService().spreadsheets().values().batchGet("").
        val map = HashMap<String, AtomicInteger>()
        val ret: MutableList<List<List<String>>> = ArrayList(ranges.size)
        for (range in ranges) {
            logger.info("range = {}", range)
            val sheetname = range.split("!").dropLastWhile { it.isEmpty() }.toTypedArray()[0]
            ret.add(
                sh.sheets.first { it.properties.title == sheetname }.data[map.computeIfAbsent(sheetname) {
                    AtomicInteger(
                        -1
                    )
                }
                    .incrementAndGet()]
                    .rowData.map {
                        it.getValues().filter { cd: CellData -> cd.effectiveValue != null }
                            .map cd@{ cd: CellData ->
                                val v = cd.effectiveValue
                                if (v.stringValue != null) return@cd v.stringValue
                                val d = v.numberValue
                                if (d % 1.0 != 0.0) return@cd String.format("%s", d) else return@cd String.format(
                                    "%.0f",
                                    d
                                )
                            }
                    }
            )
        }
        return ret

    }

    companion object {
        private val logger = LoggerFactory.getLogger(RequestGetter::class.java)
    }
}